File: frontend/src/lib/api/tag-breakdown.ts
import { apiClient } from "@/lib/api-client";
import { snakeToCamel } from "@/lib/case";
import type { TagBreakdownEntry, TagBreakdownResponse } from "@/types/tag-breakdown";

// ── Raw shapes (snake_case from backend) ──────────────────────────────────────

interface RawTagBreakdownEntry {
  tag: string | null;
  transaction_count: number;
  buy_count: number;
  sell_count: number;
  total_bought_value_by_currency: Record<string, string>;
  total_sold_value_by_currency: Record<string, string>;
}

interface RawTagBreakdownResponse {
  entries: RawTagBreakdownEntry[];
}

// ── Converters ─────────────────────────────────────────────────────────────────

function toTagBreakdownResponse(raw: RawTagBreakdownResponse): TagBreakdownResponse {
  return {
    entries: raw.entries.map(
      (entry) => snakeToCamel(entry) as unknown as TagBreakdownEntry,
    ),
  };
}

// ── Public API helpers ─────────────────────────────────────────────────────────

export async function getTagBreakdown(): Promise<TagBreakdownResponse> {
  const response = await apiClient.get<RawTagBreakdownResponse>(
    "/api/portfolio/tags/breakdown",
  );
  return toTagBreakdownResponse(response.data);
}

File: backend/app/schemas/tag_breakdown.py
"""Pydantic v2 schemas for the tag breakdown endpoint."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class TagBreakdownEntry(BaseModel):
    """Flow metrics for a single tag (or null = untagged transactions).

    Decimal monetary values are pre-serialised as strings by the service layer
    before being placed in the dict fields.
    """

    model_config = ConfigDict(from_attributes=True, str_strip_whitespace=True)

    tag: str | None = Field(
        ...,
        description="Tag string, or null for transactions with no tag",
        examples=["DCA"],
    )
    transaction_count: int = Field(
        ...,
        ge=0,
        description="Total number of transactions for this tag",
        examples=[12],
    )
    buy_count: int = Field(
        ...,
        ge=0,
        description="Number of BUY transactions",
        examples=[10],
    )
    sell_count: int = Field(
        ...,
        ge=0,
        description="Number of SELL transactions",
        examples=[2],
    )
    total_bought_value_by_currency: dict[str, str] = Field(
        default_factory=dict,
        description="Σ(qty × price) of BUY transactions per currency (Decimal → string)",
        examples=[{"USD": "1500.00", "KRW": "5000000.00"}],
    )
    total_sold_value_by_currency: dict[str, str] = Field(
        default_factory=dict,
        description="Σ(qty × price) of SELL transactions per currency (Decimal → string)",
        examples=[{"USD": "100.00", "KRW": "0.00"}],
    )


class TagBreakdownResponse(BaseModel):
    """Response envelope for GET /api/portfolio/tags/breakdown."""

    model_config = ConfigDict(from_attributes=True)

    entries: list[TagBreakdownEntry] = Field(
        default_factory=list,
        description=(
            "Per-tag flow breakdown entries. "
            "Sorted by transaction_count DESC, then tag ASC. "
            "Untagged entries (tag=null) always appear last."
        ),
    )
